### PR TITLE
Add more compiler warning output and a few more things

### PR DESCRIPTION
**Pull request by jrood-nrel**
1. Adds more compiler warning output.
2. Attempts to fix newly seen warnings.
3. Adds Werror option in CMake.
4. Enables Werror option in CI.
5. Removes hardcoded CPU core count in CI.

### CODE REVIEW

**Comment by jrood-nrel** [comment]
Actually the additional warning flags might need some logic since they're not all supported by older compilers.

File: amr-wind/physics/ConvectingTaylorVortex.cpp [hunk 1]
```diff
@@ -103,8 +103,8 @@ ConvectingTaylorVortex::ConvectingTaylorVortex(const CFDSim& sim)
     pp.query("error_log_file", m_output_fname);
     {
         amrex::Real nu;
-        amrex::ParmParse pp("transport");
-        pp.query("viscosity", nu);
+        amrex::ParmParse pp2("transport");
+        pp2.query("viscosity", nu);
```
Comment (sayerhs) [inline amr-wind/physics/ConvectingTaylorVortex.cpp:107]: In future we should just introduce a scope in the previous block. 

```c++
{
    amrex::ParmParse pp("CTV");
    pp.query("density", m_rho);
    pp.query("u0", m_u0);
    pp.query("v0", m_v0);
    pp.query("activate_pressure", m_activate_pressure);
    pp.query("error_log_file", m_output_fname);
}
```

File: cmake/set_compile_flags.cmake [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
   endif()
-  if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 7.0)
-    # Avoid notes about -faligned-new with GCC > 7
-    list(APPEND AMR_WIND_CXX_FLAGS "-faligned-new")
+  if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" OR
+     "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR
+     "${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang")
+     if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 7.0)
+       list(APPEND AMR_WIND_CXX_FLAGS "-faligned-new"
+                                      "-Wno-sign-compare"
```
Comment (sayerhs) [inline cmake/set_compile_flags.cmake:13]: Is disabling sign-compare warnings really necessary? I would prefer we kept that.

File: cmake/set_compile_flags.cmake [hunk 3]
```diff
@@ -1,20 +1,29 @@
...
+  if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
     # Intel always reports some diagnostics we don't necessarily care about
     list(APPEND AMR_WIND_CXX_FLAGS "-diag-disable:11074,11076,15335")
   endif()
-  if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 7.0)
-    # Avoid notes about -faligned-new with GCC > 7
-    list(APPEND AMR_WIND_CXX_FLAGS "-faligned-new")
+  if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" OR
+     "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR
+     "${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang")
```
Comment (sayerhs) [inline cmake/set_compile_flags.cmake:10]: Would this syntax work?
```cmake
if ("${CMAKE_CXX_COMPILER_ID}" IN_LIST "GNU;Clang;AppleClang") 
```

File: cmake/set_compile_flags.cmake [hunk 4]
```diff
@@ -1,20 +1,29 @@
...
+                                      "-Wnull-dereference"
+                                      "-Wfloat-conversion"
+                                      "-Wshadow"
+                                      "-Woverloaded-virtual")
+    endif()
   endif()
 endif()
 
+if(AMR_WIND_ENABLE_WERROR)
+  list(APPEND AMR_WIND_CXX_FLAGS "-Werror")
```
Comment (sayerhs) [inline cmake/set_compile_flags.cmake:24]: While we are at, we should also add `-fno-operator-names` here.

**Review by sayerhs — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -61,19 +57,22 @@ jobs:
     - name: Dependencies
       run: ${{matrix.install_deps}}
     - name: Setup
-      run: cmake -E make_directory ${{runner.workspace}}/build-ci-${{matrix.build_type}}
+      run: |
+        echo "NUM_PROCS=${{matrix.procs}}" >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (jrood-nrel) [reply .github/workflows/ci.yml:61]: This way shows the numerical value in the output which I prefer.

File: cmake/set_compile_flags.cmake (hunk 2 above)
Comment (jrood-nrel) [reply cmake/set_compile_flags.cmake:13]: I will try with that enabled. I really just copied AMReX's warning flags here.

File: cmake/set_compile_flags.cmake (hunk 3 above)
Comment (jrood-nrel) [reply cmake/set_compile_flags.cmake:10]: Heh, I was going to look into a generator for this, but I agree, it felt redundant. I will try that.

File: cmake/set_compile_flags.cmake (hunk 4 above)
Comment (jrood-nrel) [reply cmake/set_compile_flags.cmake:24]: Ok.

File: amr-wind/physics/ConvectingTaylorVortex.cpp (hunk 1 above)
Comment (jrood-nrel) [reply amr-wind/physics/ConvectingTaylorVortex.cpp]: Yeah this makes more sense looking at it closer.

File: cmake/set_compile_flags.cmake (hunk 3 above)
Comment (jrood-nrel) [reply cmake/set_compile_flags.cmake]: Actually this didn't seem to work. I found some code where Kitware themselves do this and it looks like this `if(NOT CMAKE_CXX_COMPILER_ID MATCHES "^(Cray|PGI|XL|XLClang)$")` . So I implemented that.

**Pull request merged**